adding clustering codes for review (#1)
magnoxemo/cardinal: doc/content/source/userobjects/ClusteringUserObject.md

> # ClusteringUserObject

Reviewer: @magnoxemo, please check out my review feedback on https://github.com/magnoxemo/cardinal/pull/4 and incorporate into this documentation.

# ClusteringUserObject

## Description

> `ClusteringUserObject` works as the base class in Cardinal for identifying clusters in mesh results, where it may be beneficial to apply

Reviewer 2: @aprilnovak  - I don't have a good feel for where this will sit in a larger documentation system, but is this place for a broader discussion of what clustering is, why it may be valauble, and how heuristics work?

mesh-tally amalgamation both in post processing or on the fly. It implements the mesh walking process. This base class assumes that only active
elements, sharing at least one common side can be clustered together based on heuristics that will be implemented 
in user objects derived from this class. 

> The `ClusteringUserObject` interacts with the AuxiliarySystem of the simulation. It iterates over 

Reviewer 2: I think this should at least be formatted, but maybe should also be a link to documentation for AuxiliarySystem?
```suggestion
The `ClusteringUserObject` interacts with the `AuxiliarySystem` of the simulation. It iterates over 
```

the mesh elements and compares their metric variable value using different heuristics.

Currently, the variable type must be `CONSTANT MONOMIAL`, since no quadrature point integration is performed
when retrieving values from the auxiliary system. If two elements are determined to
belong to the same cluster, their extra element integer IDs are unified by setting them

> to the element_id of the first element in that cluster. Any UserObject derived from `ClusteringUserObject` 

Reviewer 2: ```suggestion
to the element_id of the first element in that cluster. Any `UserObject` derived from `ClusteringUserObject` 
```

# ClusteringUserObject

## Description

> ClusteringUserObject works as the base class for clustering in Cardinal for indentifying clusters, where it could be beneficial to apply

Reviewer 2: In addition to some copyediting, I think the standard formatting for referring to this object is this:
```suggestion
`ClusteringUserObject` works as the base class in Cardinal for indentifying clusters in mesh results, where it may be beneficial to apply
```

Reviewer 2: You'll need to update each usage

Reviewer 2: @magnoxemo - you need to make this change every time you use the ClusteringUserObject name

mesh-tally amalgamation both in post processing or on the fly. ClusteringUserObject inherits from 
[GeneralUserObject](GeneralUserObject.md). It implements the mesh walking process. This base class assumes that only active

> elements, sharing at least one common side can be clustered together based on some heuristics. 

Reviewer 2: ```suggestion
elements, sharing at least one common side can be clustered together based on heuristics that will be implemented in user objects derived from this class.
```

!listing /ClusteringUserObject.h start=doco-user-object-interface-begin end=doco-user-object-interface-end include-start=false

The ClusteringUserObject interacts with the auxiliary system of the simulation. It iterates over 
the mesh elements and compares them using a specified variable, `metric_var`, which must belong to 

> the AuxVariables System. This enables the use of values computed by indicators. 

Reviewer 2: should this also be formatted like this???
```suggestion
the `AuxVariables` System. This enables the use of values computed by indicators. 
```

# ClusteringUserObject

## Description

> `ClusteringUserObject` works as the base class in Cardinal for indentifying clusters in mesh results, where it may be beneficial to apply

Reviewer: ```suggestion
`ClusteringUserObject` works as the base class in Cardinal for identifying clusters in mesh results, where it may be beneficial to apply
```

mesh-tally amalgamation both in post processing or on the fly. ClusteringUserObject inherits from 

> [GeneralUserObject](GeneralUserObject.md). It implements the mesh walking process. This base class assumes that only active

Reviewer: > ClusteringUserObject inherits from [GeneralUserObject](GeneralUserObject.md).

Is this necessary for a user to know?

elements, sharing at least one common side can be clustered together based on heuristics that will be implemented 
in user objects derived from this class. 

> !listing /ClusteringUserObject.h start=doco-user-object-interface-begin end=doco-user-object-interface-end include-start=false

Reviewer: In documentation, we don't typically show the source code unless there's something unique we are trying to show. The documentation is moreso geared towards the user (who theoretically shouldn't need to know how the source code is structured). Instead, with documentation we prefer to show snippets of input files and describe how an object is being used in an input file. I don't think this PR has any input files yet (they can come later), so I'd suggest we remove this code block altogether and make a note that when the PR into Cardinal comes later, to have an `## Examples` markdown section.

Reviewer: And because this is a base class, there wouldn't be any examples I suppose. We usually don't have documentation pages for base classes but can in this case if you anticipate this to be an actively growing system for a while. We usually relegate documentation of base classes to just Doxygen.

Author: Agreed. We can set aside those documentation pages for now. Once I submit the PR to the Cardinal main repository, I will remove this markdown file.

The ClusteringUserObject interacts with the auxiliary system of the simulation. It iterates over 
the mesh elements and compares them using a specified variable, `metric_var`, which must belong to

> the `AuxVariables` System. This enables the use of values computed by indicators.

Reviewer: I don't think it _has_ to be an auxvariable. Hypothetically it could be a variable.

> Currently, the element type must be `CONSTANT MONOMIAL`, since no quadrature point integration is performed

Reviewer: ```suggestion
Currently, the variable type must be `CONSTANT MONOMIAL`, since no quadrature point integration is performed
```
